File: android/src/main/java/com/nitroinputmask/masks/CreditCardMaskEngine.kt
package com.nitroinputmask.masks

import com.margelo.nitro.nitroinputmask.NitroMaskOptions
import com.nitroinputmask.engine.CompiledMask
import com.nitroinputmask.engine.MaskEngine
import com.nitroinputmask.engine.MaskEngineProtocol

internal class CreditCardMaskEngine(options: NitroMaskOptions) : MaskEngineProtocol {
  private val compiled: CompiledMask
  private val obfuscated: Boolean = options.obfuscated ?: false

  init {
    val mask = when (options.issuer) {
      "amex"   -> "9999 999999 99999"
      "diners" -> "9999 999999 9999"
      else     -> "9999 9999 9999 9999" // visa-or-mastercard (default)
    }
    compiled = MaskEngine.compile(mask)
  }

  override fun apply(input: String): Pair<String, String> {
    val (masked, raw) = MaskEngine.apply(input, compiled)

    if (obfuscated && masked.isNotEmpty()) {
      return Pair(obfuscate(masked), raw)
    }

    return Pair(masked, raw)
  }

  override val wantsTrailingCursor: Boolean = false
  override val expandedMask: String? = compiled.expandedMask

  private fun obfuscate(masked: String): String {
    val lastSpaceIdx = masked.lastIndexOf(' ')
    if (lastSpaceIdx == -1) return masked // single group — don't obfuscate

    val prefix = masked.substring(0, lastSpaceIdx + 1)
    val lastGroup = masked.substring(lastSpaceIdx + 1)

    val obfuscatedPrefix = prefix.map { c ->
      if (c.isDigit()) '*' else c
    }.joinToString("")

    return obfuscatedPrefix + lastGroup
  }
}

File: ios/CreditCardMaskEngine.swift
import NitroModules

struct CreditCardMaskEngine: MaskEngineProtocol {
  private let compiled: CompiledMask
  private let obfuscated: Bool

  init(options: NitroMaskOptions) {
    let issuer = options.issuer ?? "visa-or-mastercard"
    self.obfuscated = options.obfuscated ?? false

    let mask: String
    switch issuer {
    case "amex":
      mask = "9999 999999 99999"
    case "diners":
      mask = "9999 999999 9999"
    default: // visa-or-mastercard
      mask = "9999 9999 9999 9999"
    }

    self.compiled = MaskEngine.compile(mask: mask)
  }

  func apply(input: String) -> (masked: String, raw: String) {
    let (masked, raw) = MaskEngine.apply(input: input, compiled: compiled)

    if obfuscated && !masked.isEmpty {
      let obfuscatedMasked = CreditCardMaskEngine.obfuscate(masked)
      return (obfuscatedMasked, raw)
    }

    return (masked, raw)
  }

  var wantsTrailingCursor: Bool { false }
  var expandedMask: String? { compiled.expandedMask }

  // MARK: - Obfuscation

  private static func obfuscate(_ masked: String) -> String {
    guard let lastSpaceIdx = masked.lastIndex(of: " ") else {
      // No spaces — single group, don't obfuscate
      return masked
    }

    let lastGroupStart = masked.index(after: lastSpaceIdx)
    let prefix = String(masked[masked.startIndex..<lastGroupStart])
    let lastGroup = String(masked[lastGroupStart...])

    // Replace digit characters with * in prefix, keep spaces intact
    let obfuscatedPrefix = prefix.map { c -> Character in
      c.isNumber ? "*" : c
    }

    return String(obfuscatedPrefix) + lastGroup
  }
}
